### PR TITLE
feat(manifest): introduce cache layer for manifest (#16459)

### LINKED ISSUE #16459: Cache layer for manifest.
Client will pull manifest firstly when pulling an image, and harbor does not store manifest, rather proxy request to distribution, in the distribution implementation, it will read manifest from disk, in the high concurrent scenarios which will pose poor performance and high resource usage.

### PR DESCRIPTION
**Pull request by chlins**
Implement cache layer for resource manifest, it will read manifest
from cache instead of proxying to distribution if enabled.

Signed-off-by: chlins <chenyuzh@vmware.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #16459 

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/server/registry/manifest.go [hunk 1]
```diff
@@ -49,11 +52,58 @@ func getManifest(w http.ResponseWriter, req *http.Request) {
...
 
-	recorder := lib.NewResponseRecorder(w)
-	proxy.ServeHTTP(recorder, req)
+	buffer := lib.NewResponseBuffer(w)
+	// whether get manifest from cache
+	fromCache := false
+	// whether need to write-back cache
+	wbCache := false
+	// check cache
+	if config.CacheEnabled() {
```
Comment (wy65701436) [inline src/server/registry/manifest.go:61]: we have to run a OCI conformance test to validate it.

File: src/server/registry/manifest.go [hunk 2]
```diff
@@ -49,11 +52,58 @@ func getManifest(w http.ResponseWriter, req *http.Request) {
...
+		manifest, err := pkg.ManifestMgr.Get(req.Context(), art.Digest)
+		if err == nil {
+			fromCache = true
```
Comment (wy65701436) [inline src/server/registry/manifest.go:64]: did you go through this [code](https://github.com/distribution/distribution/blob/main/registry/handlers/manifests.go#L84)?

**Review by stonezdj — APPROVED** [review]

File: src/server/registry/manifest.go [hunk 3]
```diff
@@ -49,11 +62,66 @@ func getManifest(w http.ResponseWriter, req *http.Request) {
...
+	if !fromCache {
+		proxy.ServeHTTP(buffer, req)
+	}
+	// flush data
+	if _, err = buffer.Flush(); err != nil {
+		log.Errorf("failed to flush: %v", err)
+		return
+	}
+	// return if not success
+	if !buffer.Success() {
```
Comment (stonezdj) [inline src/server/registry/manifest.go:113]: We should exclude the case when the HTTP code is not 200, such as 206.

File: src/server/registry/manifest.go [hunk 4]
```diff
@@ -49,11 +62,66 @@ func getManifest(w http.ResponseWriter, req *http.Request) {
...
+		return
+	}
+	// write back manifest to cache if needed
+	if wbCache {
```
Comment (stonezdj) [inline src/server/registry/manifest.go:117]: We should exclude the case when the HTTP code is not 200, such as 206. buffer.Success() is not enough.

Comment (chlins) [reply src/server/registry/manifest.go:117]: As described in distribution spec, the success HTTP code is 200, ref: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints

**Review by wy65701436 — APPROVED** [review]
approve

**Pull request merged**